File: python/plus_one.py
class Solution:
    # @param digits, a list of integer digits
    # @return a list of integer digits
    def plusOne(self, digits):
        """
        if the lowest digit is 9, iterate make digits as 0 until we get a non 9.
        If we reach the head, insert 1 before the list.
        """
        if digits[-1] == 9:
            found_non_9 = False
            for i in reversed(range(len(digits))):
                if digits[i] == 9:
                    digits[i] = 0
                else:
                    digits[i] += 1
                    found_non_9 = True
                    break
            if not found_non_9:
                digits.insert(0, 1)
        else:
            digits[-1] += 1
        return digits


if __name__ == '__main__':
    t = [9, 9, 9, 9, 9]

    print Solution().plusOne(t)

File: c++/plus_one.cpp
#include <vector>
#include <iostream>
using namespace std;

void p(vector<int>& v) {
  for (int i = 0; i < v.size(); i++) cout << v[i] << ",";
  cout << endl;
}

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        if (digits.empty()) return digits;
        int l = digits.size();
        int carry = 1;
        for (int i = l-1; i >=0; i--) {
            int r = digits[i] + carry;
            digits[i] = r%10;
            carry = r/10;
            cout << "r:" << r << " carry:" << carry << endl;
            if (carry == 0) break;
        }
        if (carry > 0) digits.insert(digits.begin(), carry);
        return digits;
    }
};


int main() {
    Solution *s = new Solution();
    int test[] = {8,9,4,5,0,0,7,9};
    vector<int> v(test, test + sizeof(test) / sizeof(int));
    vector<int> r; 
    r = s->plusOne(v);
    p(r);
    return 0;
}
